test/irmin-pack/test_pack.ml

```diff
@@ -131,6 +133,20 @@ module Dict = struct
     ]
 end
 
+module Threads : sig
+  val with_thread : (unit -> unit) -> unit
+
+  val await : Index.async -> unit
+end = struct
+  let catch_exn = ref []
+
+  let with_thread f = try f () with exn -> catch_exn := exn :: !catch_exn
+
+  let await thread =
+    Index.await thread;
+    match !catch_exn with [] -> () | e :: _ -> raise e
+end
```

Comment:
This module confuses me :slightly_smiling_face: `with_thread` doesn't seem to be creating a thread? When we pass this into the log hook, do we not want to re-raise the exception? Is it possible for more than one exception to be accumulated in the list?

Reply:
I copied this module from index without too much thought, and I agree that the names are not great. We can rename `with_thread` to `in_thread`? 
About the exception handling part, I think the motivation was not to lose exceptions generated in the merge thread. Do you think it is not necessary?

Reply:
> About the exception handling part, I think the motivation was not to lose exceptions generated in the merge thread. Do you think it is not necessary?

It seems like a good idea to fail if a merge thread raises an exception. However, I think this module changes the semantics of the merge thread by causing it _not_ to fail if an exception is thrown: I think we probably want to re-raise the exception after logging it here, so that the thread still fails.

If so, it's (probably ?) the case that we can only ever get at most one exception in the `catch_exn` ref, so it could be a `exception option ref` instead (with an `assert false` if `Some` is detected on update). I think we can't rely on this with multiple asynchronous threads anyway, since the `cons` isn't atomic.

> We can rename `with_thread` to `in_thread`?

Maybe `with_logged_exceptions`? I don't think it has much to do with threads.

Reply:
Thanks, renamed and changed to a catching a single exception.